### PR TITLE
Add ActionReplay cheat function

### PR DESCRIPTION
**Pull request by OFFTKP**
The idea is that you add cheats through the UI, convert them to an array of uint32_t, and run each activated cheat using se_run_ar_cheat each frame.

### CODE REVIEW

**Comment by OFFTKP** [comment]
Things missing:
- Disable input when the edit/add cheat window is focused
- Better edit/add window layout
- Actually run the valid cheats from the cheats array
- Allow for pasting from the clipboard in the input inside the edit/add cheat window

**Comment by skylersaleh** [comment]
It would be cool if we added in reading/writing of a cheat database too

**Comment by OFFTKP** [comment]
Moved the AR code in nds.h (because it's different for every console) and also added the nds9_debug_read/write functions back, lmk if that's fine.

File: src/gba.h [hunk 1]
```diff
@@ -2924,6 +2925,338 @@ static uint8_t gba_audio_process_byte_write(gba_t *gba, uint32_t addr, uint8_t v
...
+          // AR slowdown
+          // Probably has no effect on emulators
+          break;
+        }
+        case 0x18:
+        case 0x1A:
+        case 0x1C:
+        case 0x1E:{
+          uint32_t address=0x8000000|((right&0xFFFFFF)<<1);
+          uint64_t decrypted=gba_decrypt_arv3(buffer[i+3] | ((uint64_t)buffer[i+2] << 32));
```
Comment (skylersaleh) [inline src/gba.h:3190]: Can you add a buffer over run check here.

File: src/gba.h [hunk 2]
```diff
@@ -2924,6 +2925,338 @@ static uint8_t gba_audio_process_byte_write(gba_t *gba, uint32_t addr, uint8_t v
...
+          gba->mem.cart_rom[address&0x1FFFFFF]=data;
+          gba->mem.cart_rom[(address+1)&0x1FFFFFF]=data>>8;
+          i+=2;
+          break;
+        }
+        case 0x10:{
+          // IF AR_BUTTON THEN [a0aaaaa]=zz
+          // Let's treat the AR button as always pressed for now
+          uint32_t address=((right<<4)&0x0F000000) | (right&0x000FFFFF);
+          uint8_t data = buffer[i+2];
```
Comment (skylersaleh) [inline src/gba.h:3201]: I think 0x10-0x14 need it for buffer[i+2] here too

File: src/gba.h [hunk 3]
```diff
@@ -2924,6 +2925,338 @@ static uint8_t gba_audio_process_byte_write(gba_t *gba, uint32_t addr, uint8_t v
...
+          break;
+        }
+        case 0x80:
+        case 0x82:
+        case 0x84:{
+          // 00000000 8naaaaaa 000000yy ssccssss  repeat cc times [a0aaaaa]=yy
+          // (with yy=yy+ss, a0aaaaa=a0aaaaa+ssss after each step)
+          uint32_t address=((right<<4)&0x0F000000) | (right&0x000FFFFF);
+          uint8_t repeat=(buffer[i+3]>>16)&0xFF;
+          uint8_t data_increment=(buffer[i+3]>>24)&0xFF;
```
Comment (skylersaleh) [inline src/gba.h:3226]: Here too

File: src/gba.h [hunk 4]
```diff
@@ -2924,6 +2925,338 @@ static uint8_t gba_audio_process_byte_write(gba_t *gba, uint32_t addr, uint8_t v
...
+          gba_store16(gba,address+offset*2,data);
+          break;
+        }
+        case 0x44:{
+          uint32_t data=right;
+          address=gba_read32(gba,address);
+          gba_store32(gba,address,data);
+          break;
+        }
+        case 0x80:{
```
Comment (skylersaleh) [inline src/gba.h:3081]: Could you add a gs_button boolean variable at the top of the function and conditionals to implement the correct operation for these (and just hard code gs_button to true). That way we don't forget about the hack that we have here.

Comment (skylersaleh) [reply src/gba.h:3081]: Same for all the other GS_button cases

**Review by skylersaleh — CHANGES_REQUESTED** [review]
Hey could you fix up these items?

**Comment by OFFTKP** [comment]
> Hey could you fix up these items?

Done

**Pull request merged**